invocation.md

```diff
@@ -0,0 +1,1170 @@
+  subject: "hello",
+  body: "world",
+})
+```
+
+## 3.1 Schema
+
+```ipldsch
+type Invocation<In> struct {
+  v      SemVer
+
+  iss    Principal
+  aud    Principal
+
+  with   URI
+  do     Ability
+  input  In (implicit {})
+
+  meta   {String : Any} (implicit {})
+
+  prf    [&UCAN]
+  nnc    optional String
+  nbf    optional Int
+
+  s      Varsig
+}
+```
+
+## 3.2 Fields
+
+An Invocation authorizes execution of the capability. There are a few invariants that MUST hold between the `iss`, `aud`, `prf` and `sig` fields:
+
+- The `iss` MUST be the Invoker.
+- All of the `prf` UCANs MUST list the Invoker in their `iss`.
+- All of the `prf` UCANs MUST list the Executor in their `aud` field, grating it authority to perform some action on a resource, or be the root authority for it.
+- The `sig` field MUST be produced by the Invoker
+- Invocation MUST be provably authorized by the UCANs in the `prf` field.
+
+### 3.2.1 UCAN Invocation Version
+
+The `v` field MUST contain the SemVer-formatted version of the UCAN Invocation Specification that this struct conforms to.
+
+### 3.2.2 Invoker
+
+The `iss` field MUST be a principal authorizing the invocation. It MUST be encoded in format describe in [UCAN-IPLD].
+
+### 3.2.3 Executor
+
+The `aud` field MUST be a principal authorized to perform the invocation. It MUST be encoded in format described in [UCAN-IPLD].
+
+### 3.2.4 Resource
+
+The `with` field MUST contain the [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) of the resource being accessed. If the resource being accessed is some static data, it is RECOMMENDED to reference it by the [`data`](https://en.wikipedia.org/wiki/Data_URI_scheme), [`ipfs`](https://docs.ipfs.tech/how-to/address-ipfs-on-web/#native-urls), or [`magnet`](https://en.wikipedia.org/wiki/Magnet_URI_scheme) URI schemes.
+
+### 3.2.5 Ability
+
+The `do` field MUST contain a [UCAN Ability](https://github.com/ucan-wg/spec/#23-ability). This field can be thought of as the message or trait being sent to the resource.
+
+### 3.2.6 Input
+
+The `input` field MUST contain any arguments expected by the URI/Ability pair. This MAY be different between different URIs and Abilities, and is thus left to the executor to define the shape of this data.
+
+UCAN capability provided in proofs MAY impose certain constraint on the type of `Input` allowed.
+
+### 3.2.7 Proofs
+
+The `prf` field MUST contain links to any UCANs that provide the authority to perform the invocation. All of the outermost proofs MUST either
+
+1. Set `aud` fields to the [Executor]'s DID and `iss` field set to the [Invoker]'s DID, allowing Executor to (re)delegate enclosed capabilities.
+2. Set `aud` field to the [Invoker]'s DID, preventing Executor from (re)delegating enclosed capabilities.
+
+### 3.2.8 Nonce
+
+If present, the OPTIONAL `nnc` field MAY include a random nonce expressed in ASCII. This field can ensures that multiple invocations are unique.
```

Comment:
if it's present but non-ascii, does that make it an invalid invocation? (before sig checks)

Reply:
I don't know what's the motivation here as it was inherited from @expede's spec. If I were doing it from scratch I'd just made `nnc` bytes.